### PR TITLE
Saisiddu/merge runon to v1

### PR DESCRIPTION
**Pull request by saisiddu**
This pull request addresses the following:
- updating the components Radiation, PotentialEvapotranpiration, SoilMoisture, Vegetation, and VegCA. 
- committed new components ResourceRedistribution and SpatialDisturbance.

Once the pull request is merged, I will request for the creation of a new release that will be included in the Landlab ecohydrology paper.

### CODE REVIEW

**Comment by saisiddu** [comment]
@mcflugen @kbarnhart or anybody: It looks like 'test_corners_at_cell' and 'test_faces_at_cell' @ C:\projects\landlab\landlab\graph\radial\tests\test_dual_radial.py are failing (from the report from appveyor on Py3_7). I updated my branch from v1, edited ecohydro components, and have pulled from v1 again before creating the pull_request. Did I mess any of these files? Can you please help me address these failures? Thank you!

**Comment by mcflugen** [comment]
@saisiddu We probably no longer need to support Python 2.7, even on landlab v1, so I think it would be fine if you just removed those builds.

The other issues look to be fairly minor. There seems to be an issue with what the starting element is when ordering elements in a rotational sense. I don't think it's a big issue, particularly because this is on the *v1* branch. These errors appear on the current version of the *v1* branch so it's not something that you've broken.

**Comment by saisiddu** [comment]
@mcflugen Thank you! In that case, I would like to merge once someone approves the changes.

**Comment by kbarnhart** [comment]
@saisiddu I have a couple of concerns about documentation and tests. 

I am not able to undertake a thorough review now, so I think it would be best if we found someone else who could do this. I should be able to do this by the end of this week if necessary. However, there are a few things that I noticed while doing a quick look through the code that I wanted to raise. 

1) Not all the component inputs are documented in the docstring (e.g., in the PotentialEvapotranspiration and SoilMoisture component init). I think its really important that these are all documented. 
2) There are not that many tests, and its unclear to me whether tests ensuring component behavior exist. Perhaps I am mistaken, but I suspect a few more are needed. I saw that the docstrings have some examples that serve as tests, and that there are unit tests, but they are primarily of the interface (e.g., ensuring that component metadata exists). I raise this point especially because you have a couple of components which have flags indicating different methods. I think that the tests should cover those different methods. Since coverage isn't building on these (since the didn't succeed), its hard to tell how well covered these components, are. I might consider running coverage locally to test.
3) Both the ResourceRedistribution and SoilMoisture components have examples, but it was really hard for me to follow what the examples were of doing.

**Comment by mcflugen** [comment]
@saisiddu I would rather not merge anything without all of the tests passing, even if it is into to the *v1* branch.

As part of this pull request, could you:
- [ ] remove the Python 2.7 builds
- [ ] fix the failing doctests (I think all that needs to be done is to change the expected output, but I can help with that)

Once the tests are passing, it will be easier to review.

**Comment by saisiddu** [comment]
Thanks, @kbarnhart!
1) I will look for the overlooked inputs in PotentialEvapotranspiration & SoilMoisture and add them.
2) I will add tests for components that have different methods (Especially PotentialEvapotranspiration & SoilMoisture). That is a good catch. For most other cases, I added doctests that actually test if the component works. I agree that they can be improved but I might have to work on them at a later time.
3) I can understand why the examples might have been difficult to follow. I added extensive examples for the ResourceRedistribution component. However, I expect the user to refer the publication to actually learn about the processes implemented in the component. I don't intend someone to completely perceive the functioning of the CA model through my component's documentation. I might be able to add extensive documentation after my dissertation is submitted.

**Comment by saisiddu** [comment]
Thanks, @mcflugen.

- I tried to fix the failing tests in test_dual_radial.py. I don't understand why the assertions failed. I ended up getting the output listed in the tests. Can you please look at them too?

**Comment by saisiddu** [comment]
@mcflugen, @kbarnhart: I uninstalled and reinstalled Landlab in the landlab-dev enviornment. Tests are running on my machine (PY3.7). I am not sure what is the issue here. Can one of you check? How do we run these tests again on appveyor and travis-ci?

**Comment by saisiddu** [comment]
@mcflugen : Travis builds failed at the linting step (it says pip couldn't install flake8). Is there something I could do to fix that? Can we skip that step so that docs can be tested?

@kbarnhart:
I added doctests to cover alternative methods (in SoilMoisture and PotentialEvapotranspiration). Local coverage shows that all ecohydro components have a mean coverage of ~90% except SoilMoisture. I suspect that my tests don't cover all 'if then else' blocks.

I would like to merge at this point and add more tests when the paper is in review.

**Comment by mcflugen** [comment]
@saisiddu You'll need to fix all of the errors reported in the linting stage first. If you run `make lint`, you can run this step locally. Running the `black` command (e.g. `black landlab/components/pet/potential_evapotranspiration_field.py`) on the files will probably get you most of the way to eliminating the errors.

**Comment by saisiddu** [comment]
Thanks @mcflugen.

**Comment by saisiddu** [comment]
@mcflugen: Do I have to install some package to run lint locally? I am on windows10.

**Comment by mcflugen** [comment]
@saisiddu Yes, you need to install *flake8* to run the linter and *black* for the reformatting. The following will install both,
```bash
$ pip install flake8 black
```

**Comment by saisiddu** [comment]
@mcflugen: I was able to run 'black' command and flake8 is already installed. But I wasn't able to run 'make lint'. I couldn't find anything online yet. I tried running 'make lint' from both landlab and landlab\landlab folders.

**Comment by mcflugen** [comment]
@saisiddu Try running the following from the top-level folder (the one with *setup.py*),
```bash
$ flake8 landlab --exclude=examples
```

**Comment by saisiddu** [comment]
@mcflugen: Thanks, Eric. I ran flake8 for each of the ecohdyro components.

I will run the command you suggested from top folder now.

**Comment by kbarnhart** [comment]
@saisiddu nice work improving the documentation and tests. 🎉 

1. A minor comment: I think you need to remove all of these files under .idea/
2. I think that it is worth having someone besides me do a review (not entirely certain if @mcflugen has looked at the code thoroughly, or was waiting for things to pass). I've added getting someone else to commit to doing a review soon to our call agenda for today. 

A side comment for @mcflugen: when I was working with @DavidLitwin he also had issues with make lint and make pretty from the command line (on a PC). I suspect its because make doesn't work out of the getgo on PC (this is speculation). Intended here as a note that we shouldn't expect those instructions to work for PC...

**Comment by DavidLitwin** [comment]
@kbarnhart I should add that the resolution for PC was to use `black landlab`, but when I tried this recently, it wanted to change additional things that seem to pass `make lint`. Not sure why this is.

**Comment by mcflugen** [comment]
@kbarnhart, @saisiddu, @DavidLitwin The problem with `make lint`, `make pretty`, etc. not working on Windows is that the *make* is a Unix command. I'm sure there's some way to install it on Windows but it's probably easier to just run the underlying commands (i.e. `flake8`, `black`, `isort`, etc.).

**Comment by kbarnhart** [comment]
We just discussed this on the call and @mcflugen noted that 
a. he was waiting for the test to pass to review and
b. he will review this by the end of the week.

**Comment by saisiddu** [comment]
Thanks, @kbarnhart, and @mcflugen.

I will remove those files under .idea (pycharm files).

File: landlab/components/plant_competition_ca/plant_competition_ca.py [hunk 1]
```diff
@@ -315,31 +347,44 @@ def update(self, time_elapsed=1, Edit_VegCov=True):
...
+            self.grid['cell']['vegetation__boolean_vegetated'] = np.zeros(
                 self.grid.number_of_cells, dtype=int
             )
-            self.grid["cell"]["vegetation__boolean_vegetated"][
+            self.grid['cell']['vegetation__boolean_vegetated'][
                 self._VegType != BARE
             ] = 1
 
         # For debugging purposes
+#        self._n = n
```
Comment (mcflugen) [inline landlab/components/plant_competition_ca/plant_competition_ca.py]: Remove commented-out code

File: landlab/components/resource_redistribution/resource_redistribution.py [hunk 2]
```diff
@@ -0,0 +1,738 @@
...
+        gr_mor_ws_thresh: float, optional
+            Background mortality probability of grass due
+            to waterstress (-)
+        """
+
+        super(ResourceRedistribution, self).__init__(grid, **kwds)
+
+        name = "vegetation__plant_functional_type"
+        if name not in self.grid.at_cell:
+            print(
```
Comment (mcflugen) [inline landlab/components/resource_redistribution/resource_redistribution.py]: Use *warning.warn* or, if you don't like that, print to *stderr*.

File: landlab/components/resource_redistribution/resource_redistribution.py [hunk 3]
```diff
@@ -0,0 +1,738 @@
...
+            sh_mor_dis_low_slp=sh_mor_dis_low_slp,
+            sh_mor_dis_high_slp=sh_mor_dis_high_slp,
+            P_sh_fire_mor=P_sh_fire_mor,
+            P_gr_fire_mor=P_gr_fire_mor,
+            sh_mor_ws_thresh=sh_mor_ws_thresh,
+            gr_mor_ws_thresh=gr_mor_ws_thresh,
+            **kwds,
+        )
+
+    def initialize(
```
Comment (mcflugen) [inline landlab/components/resource_redistribution/resource_redistribution.py]: This method seems unnecessary? Why not just initialize these variables in *__init__*?

File: landlab/components/resource_redistribution/tests/test_resource_distribution.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+"""
```
Comment (mcflugen) [inline landlab/components/resource_redistribution/tests/test_resource_distribution.py]: Could you please add more extensive tests that better test the functioning of the component?

File: landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Jan 17 11:48:13 2017
+
+@author: Sai Nudurupati & Erkan Istanbulluoglu
+"""
+
+import numpy as np
+from landlab.components import FlowRouter  # , DepressionFinderAndRouter
```
Comment (mcflugen) [inline landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py]: Remove commented-out code.

File: landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+
+def get_ordered_cells_for_soil_moisture(grid, outlet_id=None):
+    """
+    Runs Landlab's FlowRouter and DepressionFinderAndRouter to
+    route flow. Also orders the cells in the descending order of
+    channel length (upstream cell order).
+
+    Parameters:
+    ==========
```
Comment (mcflugen) [inline landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py]: Use *-* for underline, not *=*. Remove the colon.

File: landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    grid: grid object
+        RasterModelGrid
+    outlet_id: int (Optional)
+        Outlet id to be set
+
+    Returns:
```
Comment (mcflugen) [inline landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py]: Same as above.

File: landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    =======
+    ordered_cells: np.array(dtype=int)
+        cells ordered in descending order of channel length
+    grid: grid object
+        updated RasterModelGrid
+    """
+
+    if outlet_id is None:
+        outlet_id = np.argmin(grid.at_node['topographic__elevation'])
+    _ = grid.set_watershed_boundary_condition_outlet_id(
```
Comment (mcflugen) [inline landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py]: Remove the underscore.

File: landlab/components/spatial_disturbance/tests/test_spatial_disturbance.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
+"""
```
Comment (mcflugen) [inline landlab/components/spatial_disturbance/tests/test_spatial_disturbance.py]: Could you please add better tests of the component?

File: landlab/components/spatial_disturbance/spatial_disturbance.py [hunk 10]
```diff
@@ -0,0 +1,527 @@
...
+        ----------
+        grid: RasterModelGrid
+            grid, Landlab's RasterModelGrid object
+        pft_scheme: str
+            Vegetation Plant Functional Type (PFT) scheme;
+            Either 'ravi_et_al_2009' or 'zhou_et_al_2013'.
+        """
+        self._pft_scheme = pft_scheme
+        _assert_pft_scheme_is_valid(self._pft_scheme)
+        #        super(SpatialDisturbance, self).__init__(grid, **kwds)
```
Comment (mcflugen) [inline landlab/components/spatial_disturbance/spatial_disturbance.py]: Remove commented-out code.

File: landlab/components/resource_redistribution/resource_redistribution.py (hunk 3 above)
Comment (saisiddu) [reply landlab/components/resource_redistribution/resource_redistribution.py]: This is intentional. The reason I have another method is that I would like to keep the option of re-initializing the inputs (probably a subset of them) and use the same object to do so.

**Comment by saisiddu** [comment]
@mcflugen:
- I prefer not changing the case of the variables for this version. This can be taken up for version 2. I have a lot of drivers (that have contributed to the results for the paper) that are using this interface and I don't think I can deal with the changes that might be required.

- Which notebooks are you referring to? Tutorials? There are tutorials but I don't think they need to be edited for these commits (as the interface hasn't changed). I will have to add new tutorials for the new components (ResourceRedistribution and SpatialDisturbace) but it will have to wait.

- Do you suggest I move some doctests to the unit tests?

File: landlab/components/spatial_disturbance/tests/test_spatial_disturbance.py (hunk 9 above)
Comment (saisiddu) [reply landlab/components/spatial_disturbance/tests/test_spatial_disturbance.py]: Can you please give me more pointers? What do you mean by better tests?

**Comment by mcflugen** [comment]
@saisiddu 
* I was just wondering if you had written any Jupyter Notebooks for the new components. If so, we would have to move them into this repository (from, maybe, the tutorials repository) when we merge all of this into landlab 2.
* Don't forget to remove the `.idea` folder.
* You may want to move some parts of your *doctests* into unit tests. It seems like the unit tests for the new components don't really test if the components are working properly but, rather, just general *landlab* component tests (e.g. if the component has a name). When we move these components to landlab 2 it will be important to make sure things are still working and that we didn't break anything in the move. As such, lots of good tests will be extremely helpful.

File: landlab/components/plant_competition_ca/plant_competition_ca.py (hunk 1 above)
Comment (saisiddu) [reply landlab/components/plant_competition_ca/plant_competition_ca.py]: done

File: landlab/components/resource_redistribution/resource_redistribution.py (hunk 2 above)
Comment (saisiddu) [reply landlab/components/resource_redistribution/resource_redistribution.py]: done

File: landlab/components/resource_redistribution/tests/test_resource_distribution.py (hunk 4 above)
Comment (saisiddu) [reply landlab/components/resource_redistribution/tests/test_resource_distribution.py]: done

File: landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py (hunk 6 above)
Comment (saisiddu) [reply landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py]: done

File: landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py (hunk 7 above)
Comment (saisiddu) [reply landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py]: done

File: landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py (hunk 8 above)
Comment (saisiddu) [reply landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py]: done

File: landlab/components/spatial_disturbance/tests/test_spatial_disturbance.py (hunk 9 above)
Comment (saisiddu) [reply landlab/components/spatial_disturbance/tests/test_spatial_disturbance.py]: done

File: landlab/components/spatial_disturbance/spatial_disturbance.py (hunk 10 above)
Comment (saisiddu) [reply landlab/components/spatial_disturbance/spatial_disturbance.py]: done

File: landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py (hunk 5 above)
Comment (saisiddu) [reply landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py]: done

File: landlab/components/plant_competition_ca/plant_competition_ca.py (hunk 1 above)
Comment (saisiddu) [reply landlab/components/plant_competition_ca/plant_competition_ca.py]: done

**Comment by saisiddu** [comment]
Thanks, @mcflugen! I have addressed your comments, added unit tests for the new components (that test the entire component). I did remove `.idea` folder. Is it still around? 

@mcflugen, @kbarnhart : I am not sure why one of the Travis builds is failing while the other passed. Can you please look at this?

**Comment by mcflugen** [comment]
@saisiddu looks good. It appears the *.idea* folder is still there, though. Could you remove that?

**Comment by saisiddu** [comment]
Thanks a lot, @mcflugen! Merging this.

**Pull request merged**